sdk/core/azure-core/inc/azure/core/context.hpp

```diff
@@ -184,6 +186,29 @@ namespace Azure { namespace Core {
       //  That implies we return T* and NOT a T&
     }
 
+    /**
+     * @brief Get the contained value, returns \p val if value is absent.
+     * @param val A value to return when no value is contained.
+     * @return A contained value (when present), or \p val.
+     */
+    template <class T> const T& ValueOr(Key const& key, T&& val) const
```

Comment:
Can this be implemented as a convenience wrapper on top of `TryGetValue()` from https://github.com/Azure/azure-sdk-for-cpp/pull/2037

@JeffreyRichter what do you think?

Reply:
Should it be `GetValueOr()`, since we have `GetDeadline()`?

Reply:
Yes, this should be a convenience wrapper over TryGetValue. It could even be inline in the header file.

Reply:
There a problem in this function. `T&& val` can be a rvalue(a temporary object). If you return `const T&`, the return value can reference a temporary object, which will be destructed after this function returns.

So the following code will be broken.

```C++
const auto& val = context.ValueOr("NotExistKey", FunctionReturnsRValue());
```

Reply:
My suggestion is to add the following line into the function

```C++
static_assert(!std::is_rvalue_reference(val));
```

so that the user can get a build error if he passes an rvalue as 2nd parameter.